File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int sum=0;
        int max_sum=INT_MIN;
        int total_sum=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i%nums.size()];
            total_sum+=nums[i%nums.size()];
            if(sum>max_sum){
                max_sum=sum;
            }
            if(sum<0){
                sum=0;
            }


        }
        sum=0;
        int min_sum=INT_MAX;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i%nums.size()];
            if(sum<min_sum){
                min_sum=sum;
            }
            if(sum>0){
                sum=0;
            }
        }
        if(max_sum<0){
            return max_sum;
        }
        return max(max_sum,total_sum-min_sum);
        
        
        
    }
};

File: 954-maximum-sum-circular-subarray/maximum-sum-circular-subarray.java
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        int max_sum=nums[0];
        int min_sum=nums[0];
        int curr_sum=nums[0];
        int curr_min=nums[0];
        int total=0;
        for(int i=1;i<nums.length;i++){
            curr_sum=Math.max(nums[i],curr_sum+nums[i]);
            max_sum=Math.max(max_sum,curr_sum);
            curr_min=Math.min(nums[i],curr_min+nums[i]);
            min_sum=Math.min(min_sum,curr_min);
            total+=nums[i];
            
        }
        total+=nums[0];

        return max_sum<0?max_sum:Math.max(max_sum,total-min_sum);
        
    }
}